pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go

```diff
@@ -0,0 +1,762 @@
+			return true
+		}
+		return objects[i].GetName() < objects[j].GetName()
+	})
+}
+
+type namespacedName struct {
+	Name      string `json:"name"`
+	Namespace string `json:"namespace"`
+}
+
+type object struct {
+	Metadata namespacedName `json:"metadata"`
+}
+
+func update[T metav1.Object](t *testing.T, objects []T, updates []any) []T {
+	var updated []T
+
+	// Convert all updates from apply configuration to the corresponding
+	// JSON. By decoding the JSON into namespacedName we can figure out
+	// what the update is for.
+	jsonUpdates := make(map[namespacedName][]byte)
+	for _, update := range updates {
+		data, err := json.Marshal(update)
+		require.NoError(t, err)
+		var o object
+		require.NoError(t, json.Unmarshal(data, &o))
+		jsonUpdates[o.Metadata] = data
+	}
+
+	for _, object := range objects {
+		// We can get this through the common interface.
+		id := namespacedName{
+			Name:      object.GetName(),
+			Namespace: object.GetNamespace(),
+		}
+
+		if data, ok := jsonUpdates[id]; ok {
+			// Updates are not proper server-side-apply - here we just overwrite fields.
+			require.NoError(t, json.Unmarshal(data, object))
+		}
+		updated = append(updated, object)
+	}
+	return updated
+}
+
+func setup(t *testing.T, nodes []*v1.Node, claims []*resourcev1alpha1.ResourceClaim, classes []*resourcev1alpha1.ResourceClass, schedulings []*resourcev1alpha1.PodScheduling) (result *testContext) {
+	t.Helper()
+
+	tc := &testContext{}
+	_, ctx := ktesting.NewTestContext(t)
+	ctx, cancel := context.WithCancel(ctx)
+	t.Cleanup(cancel)
+	tc.ctx = ctx
+
+	tc.client = fake.NewSimpleClientset()
+	reactor := createReactor(tc.client.Tracker())
+	tc.client.PrependReactor("*", "*", reactor)
+
+	informerFactory := informers.NewSharedInformerFactory(tc.client, 0)
+
+	opts := []runtime.Option{
+		runtime.WithClientSet(tc.client),
+		runtime.WithInformerFactory(informerFactory),
+	}
+	fh, err := runtime.NewFramework(nil, nil, tc.ctx.Done(), opts...)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	pl, err := New(nil, fh, feature.Features{EnableDynamicResourceAllocation: true})
+	if err != nil {
+		t.Fatal(err)
+	}
+	tc.p = pl.(*dynamicResources)
+
+	// The tests use the API to create the objects because then reactors
+	// get triggered.
+	for _, claim := range claims {
+		_, err := tc.client.ResourceV1alpha1().ResourceClaims(claim.Namespace).Create(tc.ctx, claim, metav1.CreateOptions{})
+		require.NoError(t, err, "create resource claim")
+	}
+	for _, class := range classes {
+		_, err := tc.client.ResourceV1alpha1().ResourceClasses().Create(tc.ctx, class, metav1.CreateOptions{})
+		require.NoError(t, err, "create resource class")
+	}
+	for _, scheduling := range schedulings {
+		_, err := tc.client.ResourceV1alpha1().PodSchedulings(scheduling.Namespace).Create(tc.ctx, scheduling, metav1.CreateOptions{})
+		require.NoError(t, err, "create pod scheduling")
+	}
+
+	informerFactory.Start(tc.ctx.Done())
+	t.Cleanup(func() {
+		cancel()
+		informerFactory.Shutdown()
+	})
```

Comment:
you are already cleaning the context above

Reply:
Cancel is called above as soon as the context exists, just in case that the setup fails halfway through. It needs to be called again here before `informerFactory.Shutdown()` because otherwise the goroutines that we wait here wouldn't stop (t.Cleanup invokes first-in-last-out).

I added some comments explaining this.